src/FilesystemStore.js

```diff
@@ -0,0 +1,181 @@
+'use strict';
+
+const fs = require('fs').promises;
+const { join } = require('path');
+const jsonl = require('./jsonl');
+const { writeStrings, readStrings } = require('./strings');
+
+/**
+ * @typedef {Object} TelegramMessagesFilter
+ *
+ * @property {Number} [unix_date] Date the message was sent in Unix Time format.
+ * @property {String} [sender_role] Message sender's role (i.e. "admin", "user", "bot").
+ * @property {Number} [sender_id] Message sender's id.
+ * @property {String[]} [tags] Tags attached to saved message.
+ * Interpretation on this filter parameter is defined by the "tags_mode" filter parameter.
+ * @property {'all'|'any'|except} [mode] Defines the way filter parameters will be interpreted.
+ * @property {'all'|'any'|except} [tags_mode] Defines the way filter tags will be interpreted.
+ * Equals to the "mode" filter parameter if not set explicitly.
+ */
+
+/**
+ * @typedef {Object} TelegramMessage
+ *
+ * @see https://core.telegram.org/bots/api#message
+ */
+
+const filterPredicates = {
+    all: (collection, lambda) => collection.every(lambda),
+    any: (collection, lambda) => collection.some(lambda),
+    except: (collection, lambda) => !collection.some(lambda),
+};
+
+class FilesystemStore {
+    constructor(basePath, params = {}) {
+        return (async () => {
```

Comment:
To create async constructor you may return new Promise... or even better `return this.init()` while init will be async method.